docs/react-three-fiber/tutorials/typescript.mdx

```diff
@@ -0,0 +1,130 @@
+---
+title: Using with TypeScript
+description: This guide will help through common scenarios  and how to approach them with TypeScript.
+nav: 17
+---
+
+This tutorial will assume some React and TypeScript knowledge and will base off of [this starter codesandbox](https://codesandbox.io/s/brnsm), so you can fork it and follow along.
+
+## Typing refs
+
+React's `useRef` won't automatically infer types despite pointing it to a typed element.
+
+You can type the ref yourself by passing an element's type through via generics:
+
+```tsx
+import { useRef, useEffect } from 'react'
+import { Mesh } from 'three'
+
+function Box(props) {
+  const ref = useRef<Mesh>(null!)
+
+  useEffect(() => {
+    console.log(Boolean(ref.current))
+  }, [])
+
+  return (
+    <mesh {...props} ref={props}>
+      <boxGeometry />
+      <meshBasicMaterial />
+    </mesh>
+  )
+}
+```
+
+The exclamation mark is a non-null assertion that will let TS know that `ref.current` is defined when we access it in effects.
+
+## Typing shorthand props
+
+react-three-fiber accepts short-hand props like scalars, strings, and arrays so you can declaratively set properties without side effects.
+
+Here are the different variations of props:
+
+```tsx
+import { Euler, Vector3, Color } from 'three'
+
+rotation: Euler || [x, y, z]
+position: Vector3 || [x, y, z] || scalar
+color: Color || 'hotpink' || 0xffffff
+```
+
+Each property has extended types which you can pull from to type these properties.
+
+```tsx
+import { Euler, Vector3, Color } from '@react-three/fiber'
+// or
+// import { ReactThreeFiber } from '@react-three/fiber'
+// ReactThreeFiber.Euler, ReactThreeFiber.Vector3, etc.
+
+rotation: Euler
+position: Vector3
+color: Color
+```
+
+This is particularly useful if you are typing properties outside of components, such as a store or a hook.
+
+## Extend usage
+
+react-three-fiber can also accept third-party elements and extend them into its internal catalogue.
+
+```tsx
+import { useRef, useEffect } from 'react'
+import { GridHelper } from 'three'
+import { extend } from '@react-three/fiber'
+
+// Create our custom element
+class CustomElement extends GridHelper {}
+
+// Extend so the reconciler will learn about it
+extend({ CustomElement })
+```
+
+The catalogue teaches the underlying reconciler how to create fibers for these elements and treat them within the scene.
+
+You can then declaratively create custom elements with primitives, but TypeScript won't know about them nor their props.
+
+```html
+// error: 'customElement' does not exist on type 'JSX.IntrinsicElements'
+
+<customElement />
+```
```

Comment:
VSCode highlights this a bit differently:

![image](https://user-images.githubusercontent.com/23324155/144760418-e103c9c7-d79b-4557-8b82-39ca6a5dd7e0.png)

![image](https://user-images.githubusercontent.com/23324155/144760424-b3ef261a-a7b6-41a5-a93d-e8da6bc6b2e8.png)